Add Inline Images and Inline Blocks API

jayshenk opened the pull request:
# Resolves
Insert images inline #2043.

# UI Design

## Goal

The user should be able to add inline images in text, following the steps in the mockups by @jasmussen in https://github.com/WordPress/gutenberg/issues/2043#issuecomment-373301990.

## Tasks:
- [x] When the user places the caret in RichText and clicks the global inserter, an 'Inline Blocks' menu should be rendered with the 'Inline Image' button. Empty Paragraphs are an exception and the 'Inline Blocks' menu will not be rendered.
- [x] At the same time, a blue rectangle should appear by the caret position to show where the image will be inserted. The block level insertion point after the block should *not* be displayed.
- [x] When the user selects the 'Inline Image' button, the media library modal should be shown.
- [x] After an image is selected, it is added inline with the text at the caret position.

### Future Plans:
- Add @ mentions as an Inline Block that can be added with the global inserter.

# Technical Design

## Components/Modules Affected

### `Inserter`

Determines whether to render the regular blocks menu (and the regular insertion point) or Inline Blocks menu (and inline insertion point) by communicating with the editor redux store. 

### `InserterInlineMenu`

Uses the existing `InserterGroup` component to render the buttons for the registered inline blocks.

### `RichText`

When selected, renders the `InlineBlocks` component.

### `InlineBlocks`

On mount, dispatches an action to the editor store signalling that inline block insertion is available. Responsible for communicating with editor store, and inserting inline block using the TinyMCE editor instance. Renders the `InlineInsertionPoint` when editor store signals it to.

### `InlineInsertionPoint`

Absolutely positioned inline insertion point (caret and indicator) where the inline block will be inserted.

### `MediaUpload`

Used to interact with the media library to select the image for the inline image.

### `inline-blocks/`

Provides an API for registering and getting inline block types. Modeled after the regular blocks API (in `blocks/` directory) to provide a similar interface to developers.

### `core-inline-blocks/`

Stores the core inline block types, mirroring the directory structure of `core-blocks/`. Currently just has the inline image, but later could be expanded to include other inline blocks like @ mentions.

### Editor store reducers/actions/selectors

Used to communicate between the `InlineBlocks` component and the `Inserter`.

New state added:
	
`isInlineInsertAvailable` (`boolean`)
		
Set to true by the `InlineBlocks` component when `RichText` is selected. This lets the `Inserter` know that it can render the inline blocks menu.

`isInlineInsertionPointVisible` (`boolean`)
		
Set to true by the `Inserter` when it determines that we are inserting inline and renders the Inline Blocks menu. The `InlineBlocks` component subscribes to this and renders the `InlineInsertionPoint`.
	
`inlineBlockNameForInsert` (`string`)

Set by the `Inserter` when an inline block is selected from the menu. The `InlineBlocks` component subscribes to this to know which inline block to insert.

## Additional Design Choices

- Allow inline blocks in any RichText https://github.com/WordPress/gutenberg/pull/5794#pullrequestreview-113682335
- Separate Inserter menus for regular blocks and inline blocks (when new inserter ready, consider adding it to that above the other items https://github.com/WordPress/gutenberg/pull/5794#issuecomment-382684045)
- Set initial width of inline image to 150px for large images (can be resized larger or smaller after insertion)

Previously considered and alternative approaches we chose *not* to do:

- Using block `supports` to enable/disable inline blocks at the block level
- Having a prop for RichText that enables/disables inline blocks https://github.com/WordPress/gutenberg/pull/5794#pullrequestreview-113682335
- Adding insertion point indicator into the TinyMCE content https://github.com/WordPress/gutenberg/pull/5794#issuecomment-379131526
- Naming (previously called 'inline tokens' before changing to 'inline blocks' to match the UI).

### Currently working on:

https://github.com/WordPress/gutenberg/pull/5794#pullrequestreview-117814793
- Moving `MediaUpload` logic out of `InlineBlocks` component. This should be in the registered inline block or perhaps rendered by the `Inserter`.
- Solidifying the render method of the registered inline block. Allow it to return a React element, and determine if it could also return a full component with lifecycle (need to determine if TinyMCE content will allow this).

### Future plans:

- Combining inline menu with new inserter design after it's merged https://github.com/WordPress/gutenberg/pull/6636.
- Adding @ mentions as an inline block.
- Consider adding an edit toolbar for inline image https://github.com/WordPress/gutenberg/pull/5794#issuecomment-381516924.

## Screenshots (jpeg or gifs if applicable):
![inline-image-3](https://user-images.githubusercontent.com/10903492/39893027-8257d542-5470-11e8-9f67-a29291e5073f.gif)

## Tests

- E2E for inserting inline image in Paragraph
- Unit tests for reducer, actions, and selectors
- Manually tested adding inline image in all core blocks that use RichText
- Plan on adding tests of components and inline block registration after making some changes to them

jayshenk committed:
Global inserter renders Inline Blocks menu when appropriate

Inserter checks whether the currently selected block has
`canInsertTokens` attribute to determine whether Inline Blocks menu
should be rendered instead of the Blocks menu.

jasmussen commented:
Nice work, very good description and plan seems to make sense. Also nice first step. Here's a GIF:

![inline images](https://user-images.githubusercontent.com/1204802/37890040-4d8e6a84-30cf-11e8-929d-2072b6bcdf45.gif)

This is a great way to start, because it can help us validate whether the plan as laid out in #2043 indeed works or not. My initial impression is that yes, it does, but we need to be mindful of a couple of things:

- In an empty paragraph, you should see the standard blocks menu, not the inline blocks menu. Even though you're in text, this is a special case. CC: @youknowriad 
- If you put the caret in text, click the inserter, you see inline blocks. Cool! If you click outside the menu, it should immediately close — right now you very briefly see the blocks menu appearing instead (see GIF)

Good items on your checklist. Wanted to elaborate on this:

> At the same time, a blue square should appear by the cursor position to show where the image will be inserted.

Forgive me if this was implied or obvious, but the blue line that appears _after_ the block should not be present in this situation. That is the indicator for where block level block would go. 

Tagging this ticket and CC'ing a few. Nice work! What help do you need with next steps?

gziolo commented on blocks/library/paragraph/index.js:
```diff
@@ -311,6 +311,10 @@ const schema = {
 	customFontSize: {
 		type: 'number',
 	},
+	canInsertTokens: {
```
It probably shouldn't be a part of `attributes`, because their role is to help with parsing block's definition into data to consume. It should be rather inside `supports` (see https://github.com/WordPress/gutenberg/blob/master/docs/block-api.md#supports-optional) property or directly inside `settings` part of block's definition.

jayshenk replied:
Makes sense, thanks. I don't currently see a way to access the `supports` or `settings` in the editor store. Perhaps this is a sign that I need to add new Redux actions? The design discussion happening now as to where the button to add inline images should be may have some implications here as well.

aduth replied:
The store only tracks the name, UUID, and current attribute state of a block. But the name is enough to get the supports of a block, via [`getBlockSupport( name, feature, defaultSupports )`](https://github.com/WordPress/gutenberg/blob/2ae21083a8deb71685523a6de35647b774ba351f/blocks/api/registration.js#L258-L276), [`hasBlockSupport( name, feature, defaultSupports )`](https://github.com/WordPress/gutenberg/blob/2ae21083a8deb71685523a6de35647b774ba351f/blocks/api/registration.js#L278-L290), or simply [`getBlockType( name ).supports`](https://github.com/WordPress/gutenberg/blob/2ae21083a8deb71685523a6de35647b774ba351f/blocks/api/registration.js#L238-L247).

(Aside: Wondering if these ought to be called `getBlockTypeSupport`, `hasBlockTypeSupport` for terminology consistency)

jayshenk replied:
Ah, thanks for pointing out those functions! I hadn't noticed them yet and they should be useful.

karmatosed commented:
> This is a great way to start, because it can help us validate whether the plan as laid out in #2043 indeed works or not.

I also think it does work and maybe sets a nice precedent for the future. I have a few points to consider on discoverability though.  I am concerned firstly if inline will be what someone would look for, secondly if it's findable. For example, how do I know to look for inline blocks? This is a bigger question but does open it. Should the action to add maybe be 'by' the block if it's inline?

jayshenk commented:
From @jasmussen:
> - In an empty paragraph, you should see the standard blocks menu, not the inline blocks menu. Even though you're in text, this is a special case. 
> - If you put the caret in text, click the inserter, you see inline blocks. Cool! If you click outside the menu, it should immediately close — right now you very briefly see the blocks menu appearing instead (see GIF)

Noted, thanks!

>  the blue line that appears after the block should not be present in this situation. That is the indicator for where block level block would go.

Thanks, I added that to the task description.

From @karmatosed:
> Should the action to add maybe be 'by' the block if it's inline?

I could see how this might be easier for the user to find. Would it be an image button on the toolbar for the block?

mtias commented:
I'd like to avoid having this by the block as it fragments even further the ways in which you can insert content and is going to be confusing — people might expect that's _the_ way of adding any image, for example. The current plan still seems sound to me.

jayshenk committed:
Use block `supports` to determine Inserter state

jayshenk committed:
Don't show Inline menu if Paragraph content empty

jayshenk committed:
Fix failing e2e test

Need to press 'Enter' after typing in Paragraph so that the regular Inserter
menu will render instead of the Inline Blocks menu

mtias commented:
> I'm wondering if this is in the scope of this PR? Or is it ok to hard code the 'Inline Image' button in the Inserter for now, to see if we're happy with the interactions, and save the inline blocks API for a future PR?

Up to you if you want to explore it. Probably still good to have it in a separate PR for ease of reviewing and implementing.

jayshenk committed:
Must be in visual mode to insert Inline Blocks

jayshenk committed:
Add state for inline insertion point visibility

jayshenk committed:
Inserter dispatches inline insertion actions

jayshenk commented:
I'm currently working on adding the inline insertion point (the blue square that appears after the caret showing where the image will be inserted). I did a lot of experimenting with TinyMCE and adding the square by inserting an element into the RichText/TinyMCE content. This approach seems problematic, as I was running into the following problems:

- TinyMCE will remove elements like empty `span`s and `div`s from the content
- the undo manager sometimes gets confused by the addition of the element (there are workarounds to this, but it seems prone to bugs)
- there is a risk of the added element getting saved into the block content (again, it is possible to circumvent this, but it seems prone to bugs)

Because of these potential issues, I'm now thinking it may be better to have the insertion point be absolutely positioned above the RichText rather than altering the content. Do you foresee any problems with this approach @jasmussen? If I absolutely position the blue square, the square may potentially cover up a letter if the caret was in the middle of a word. Otherwise, when the caret is at the end of a word, as I would think would normally be the case when adding an inline image, the appearance would look like the mockup. If this description isn't clear, I can provide screenshots/gifs.

jasmussen commented:
> Because of these potential issues, I'm now thinking it may be better to have the insertion point be absolutely positioned above the RichText rather than altering the content. Do you foresee any problems with this approach @jasmussen?

While I don't foresee any problems with this approach, the blue insertion indicator has less value than the inline image itself. If you like, you could postpone or even ignore that aspect of the mockups, and focus on getting the inline image insertion to work instead. Totally up to you 😄

jayshenk committed:
Add InlineInsertionPoint component

jayshenk commented:
@jasmussen Thanks for that perspective! I probably spent too much time going down the TinyMCE rabbit hole... but I think the knowledge I gained about its API will be useful when I do the inline image insertion. Using absolute positioning for the blue insertion indicator was much simpler so I went ahead and added that. Next steps: rendering the media library and inserting the inline image upon selection.

jayshenk committed:
Clicking Inline Image button renders Media Library

jayshenk committed:
Add inlineToken state to editor store

jayshenk committed:
Inline image inserted in RichText upon selection

jayshenk committed:
Only insert token if RichText selected

jayshenk commented:
Happy to report that the feature is working now! I added a gif to the PR description. Still need to work out a bug, write tests, and documentation, but I would appreciate a review of what I have done so far @jasmussen.

jasmussen commented:
Really nice work! 🔥

Here's a GIF:

![inline-images](https://user-images.githubusercontent.com/1204802/38662083-6ceb598e-3e33-11e8-88ee-344c73dcd861.gif)

I can't seem to break it. This seems to work solidly well!

I quickly tested in Chrome's mobile emulator, where resizing images doesn't seem to work. But this could either be due to it being the mobile emulator not behaving and it might still work on an actual device, or it could simply be a limitation of TinyMCE. In either case addressing that is not necessary as part of this PR, and definitely something we can look at separately. (And in doing so it would be nice if we could get the big friendly blue drag handles that the image block uses for resizing). But again, a separate thing.

Stellar work! 👏

karmatosed reviewed (approved):
This is really great!

noisysocks commented:
I think that it's a little confusing how if you type some text and then try to insert e.g. a quote block with your mouse, the UI gets in your way:

![ux](https://user-images.githubusercontent.com/612155/38719289-761fa918-3f34-11e8-8312-85292c690644.gif)

(Not a blocker: we can leave this as is for now and iterate after more comprehensive user testing 🙂)

jasmussen commented:
I think that's a really solid point @noisysocks makes. And this is one of those questions that's come up under the development, where we weren't quite sure what would work until we tried it. 

We could revisit an idea originally posed on the mockups, which is to always show the blocks, but _also_ show inline blocks when in text:

![inserter](https://user-images.githubusercontent.com/1204802/38723445-6006c022-3f01-11e8-8bb2-86b6d2441bd9.png)

CC: @mtias because I think he had some thoughts about this.

jayshenk commented:
I agree that the UI concern @noisysocks raised makes sense, and I'll look forward to seeing if there is consensus around the mockup @jasmussen referred to. If we did combine the menus as in the mockup, we would need to consider what the behavior of the insertion indicator would be (for instance, would it switch between the inline indicator and the block indicator while the menu was open?).

I had a conversation with @nb today about this PR, and here's a summary of what we discussed -- would be great to get others input:

First was a concern about using the term `token` or `inlineToken` in the code to name Inline Blocks. 'Token' is rather opaque and may not be easy for other developers to understand what they are. Perhaps they should be named `inlineBlocks` to match the UI. My only concern would be whether they would be confused with standard blocks which have a different API.

Thinking ahead to when we add Inline Blocks to other blocks that use RichText, wondering whether checking a block's `supports` as I'm doing currently is the best way to determine if Inline Blocks can be added. Potentially, in a block that includes multiple RichText components, it could be that the developer only wants some of those components to allow Inline Blocks. In that case, perhaps it is better to enable/disable Inline Blocks as a prop for each RichText. We were thinking that perhaps by default Inline Blocks could be allowed in RichText, but could be disabled if needed with the prop. This has some implications for the Inserter, which is currently checking the block `supports`, but would instead need to have awareness of whether RichText was selected and whether inline blocks were enabled within it. Perhaps using the new React `context` API for `isSelected` as being worked on in https://github.com/WordPress/gutenberg/pull/5029 by @gziolo could be useful here.

This also brings up the question of whether other blocks with RichText should behave similarly to the Paragraph block, in that if the RichText is empty, the regular blocks menu should render, and only if the RichText has content should the Inline Blocks menu render.

Another question that was brought up, was if there is anything more that needs to be added to the HTML that the Inline Image generates. Currently it is just an `img` tag with `src` and `alt` attributes. Should there be additional elements or attributes to identify or parse it as an Inline Block?

afercia commented:
I think IE11 has some problems with images in a `contenteditable` element, maybe worth starting some investigation. Will open a specific issue though.

jasmussen commented:
I'll let some of the developers answer your other questions.

> This also brings up the question of whether other blocks with RichText should behave similarly to the Paragraph block, in that if the RichText is empty, the regular blocks menu should render, and only if the RichText has content should the Inline Blocks menu render.

This is a good question, and probably something we don't want to answer right now, or even in this PR. I think it would be good to get the baseline merged — as you say, render the regular blocks menu if an empty paragraph, render the inline blocks menu in all other rich text blocks and a paragraph with text. We can then revisit depending on how this feels to use for a while.

> Another question that was brought up, was if there is anything more that needs to be added to the HTML that the Inline Image generates. Currently it is just an img tag with src and alt attributes. Should there be additional elements or attributes to identify or parse it as an Inline Block?

Great catch. `src` and `alt` seems to be sufficient, considering you can resize using the handles. 

But it seems like  we need to find a way to allow the user to add an alt tag, without having to enter the code editor. Can we allow the TinyMCE Image toolbar to show up here, change the visuals to match Gutenberg toolbars, and remove the alignment buttons?

<img width="263" alt="screen shot 2018-04-16 at 10 13 10" src="https://user-images.githubusercontent.com/1204802/38797496-ed6c7b0a-415e-11e8-8936-b78a01f5061d.png">

This would give us a handy Edit button, that would open the media modal, allowing you to edit the alt text.

jayshenk committed:
WIP: Insert Inline Blocks into any RichText

jayshenk committed:
Add insertion point margin calculation

Necessary in order for insertion point to display at correct position
in list block.

jayshenk committed:
Add `inlineBlocksEnabled` prop to RichText

Prop is used to enable/disable Inline Block insertion in a RichText
component.

Disabled Inline Blocks in the Image, Gallery, Button, and Text Column
Blocks. Enabled in all other blocks that use RichText such as Quote,
Table, Verse, and List Blocks.

jayshenk committed:
Remove `inlineToken` from Paragraph `supports`

No longer needed since enabling Inline Blocks is now done by
specifying a prop for RichText.

jayshenk committed:
Rename `token` and `inlineToken` to `inlineBlock`

It is not clear what `token` means, so instead use `inlineBlock` name
to match the UI.

jayshenk committed:
Give large inline images initial width of 150

Image can be resized if it needs to be larger, but the goal here is to
provide a sensible default.

jayshenk committed:
Make inline insertion unavailable on unmount

Fixes bug where Inline Blocks menu was rendering after Paragraph was
replaced by a different block.

jayshenk committed:
Fix failing e2e test

Need to tab out of Quote block to prevent Inline Blocks menu from
rendering.

gziolo commented on editor/components/inserter/index.js:
```diff
@@ -54,34 +101,58 @@ class Inserter extends Component {
-					return <InserterMenu onSelect={ onSelect } />;
-				} }
-			/>
+			<div>
```
Is this `div` necessary? Would it work with `Fragment` component?

jayshenk replied:
The `div` is not necessary, so I changed it to a `Fragment` component as you suggested. Thanks for noticing that!

jayshenk committed:
Use Fragment instead of div

jayshenk committed:
Close Inline Blocks menu upon selection

Fixes problem on mobile, where menu remained open after Inline Image
button was selected.

jayshenk commented:
A few updates:

You can now add inline images to any RichText. Inline blocks can be enabled/disabled in RichText with the `inlineBlocksEnabled` prop. Currently I have enabled Inline Blocks in the RichText for Paragraph, Quote, List, Verse, and Table blocks. I disabled them in the Image, Gallery, Button, and Text Column blocks as those blocks would need to be modified to make them compatible.

Secondly, for images that have a width greater than 150px, I default them to a width of 150px when inserting them inline. This seemed a reasonable default, and the image can be resized to be larger or smaller as needed. It prevents a large inline image from taking over the screen or appearing like an Image Block.

Here's a GIF:
![inline-image-2](https://user-images.githubusercontent.com/10903492/38970132-b837d48e-4361-11e8-86a6-4fba224ee9d9.gif)

On a code level, I renamed `token` and `inlineToken` to `inlineBlock` to match the UI and hopefully be a more meaningful name to other developers.

--

From @jasmussen in https://github.com/WordPress/gutenberg/pull/5794#issuecomment-381056583:
> We could revisit an idea originally posed on the mockups, which is to always show the blocks, but also show inline blocks when in text

Would it look like the mockup, with no Search field or tabs like the normal inserter menu? Which of the standard blocks would be included in the menu?

https://github.com/WordPress/gutenberg/pull/5794#issuecomment-381516924:
> But it seems like we need to find a way to allow the user to add an alt tag, without having to enter the code editor. Can we allow the TinyMCE Image toolbar to show up here, change the visuals to match Gutenberg toolbars, and remove the alignment buttons? This would give us a handy Edit button, that would open the media modal, allowing you to edit the alt text.

Good point, you can add the `alt` text in the media library when first inserting the image, but there is no way to edit the `alt` attribute without editing the HTML. I'll look into adding the edit toolbar as you suggested.

jasmussen commented:
Solid thoughts, enhancements look good. 

> Would it look like the mockup, with no Search field or tabs like the normal inserter menu? Which of the standard blocks would be included in the menu?

Eagle eyed! 

It would be a version of what's ongoing in https://github.com/WordPress/gutenberg/pull/6168, but combined with the previous mockup posted. Here's a refresh that includes the search bar:

![inserter](https://user-images.githubusercontent.com/1204802/38985970-49c91746-43cb-11e8-9c92-aeb7fb0cb595.png)

However this inserter work should possibly not be part of THIS branch — we might want to keep this branch unmerged until a decision has been made on the other branch, so that can be merged in, and this can be rebased on top of that perhaps. CC: @youknowriad

iseulde commented on blocks/library/button/index.js:
```diff
@@ -98,6 +98,7 @@ class ButtonBlock extends Component {
 							color: textColor,
 						} }
 						keepPlaceholderOnFocus
+						inlineBlocksEnabled={ false }
```
Should we let blocks disable this? I think it's kind of cool that you can insert inline images in any `RichText` field. I can even imagine this having use in captions, e.g. if I want to insert a little custom GIF emoji. I think we should give the users the freedom to decide, not the block, as it is still valid. Cc @jasmussen.

jasmussen replied:
I tend to agree that this is cool. We may find that we do want to allow a block to disable this, but it seems worth starting out liberal and then reining in once a need arises.

This partially depends on how well the combined inline and block level library works.

iseulde replied:
Yeah I think it's worth to be liberal here and see what happens. Also `inlineBlocksEnabled` wouldn't work that well if we later add some other blocks that we then want here too.

jayshenk replied:
Ok I'm convinced 😃. I removed the `inlineBlocksEnabled` prop in https://github.com/WordPress/gutenberg/pull/5794/commits/51dc693ce8e6574da270e2e7e1bc4ce020e17ce5 so that users can choose to insert inline images in any RichText.

Currently the Image and Gallery blocks prevent the inline image from being resized, and set it to `display: block`, but I think I can fix these issues by adding CSS for the inline image.

The Text Column block has an issue where the inline image is duplicated into both columns. It looks like it is related to how `isSelected` is used, and similar to issue https://github.com/WordPress/gutenberg/issues/5633. Luckily, looks like this will be fixed in https://github.com/WordPress/gutenberg/pull/6280.

gziolo replied:
Yes, should be fixed now. I just merged #6280 into `master` :)

jayshenk replied:
Awesome, I rebased and inline images are working correctly in the Text Column block now. Thanks @gziolo!

nb replied:
On the surface I agree with the idea to allow the inline images everywhere, though I would be on the lookout for places where we might want only text.

also cc @mtias who had an opinion

jayshenk replied:
Makes sense. Inline images do look a bit out of place in the Gallery block:
<img width="640" alt="screen shot 2018-04-26 at 11 51 17 pm" src="https://user-images.githubusercontent.com/10903492/39343813-ed085f5c-49ac-11e8-8c15-ecdc22bf565a.png">

But perhaps there are use cases such as the custom GIF emoji as @iseulde mentioned where the user may want this option. And I can see why it may be good for the user to decide whether to insert it, rather than us deciding for them. 

Also, once user mentions are an inline block option, they may be appropriate for captions as well. But if we do come across places where we want to disable inline blocks, I'm happy to add the RichText prop back in.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -441,6 +460,60 @@ export class RichText extends Component {
 
+	getInsertionPosition() {
+		const container = this.getContainerNode();
+		const marginLeft = window.getComputedStyle( container )[ 'margin-left' ];
```
Not sure if this could ever error in any way, but it might be prudent to use e.g. Lodash's `get` here.

jayshenk replied:
I agree, that would be safer. Made the change in https://github.com/WordPress/gutenberg/pull/5794/commits/9142f64e98f8a07b6c11b8ff4dd7acf9e092ccc0.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -441,6 +460,60 @@ export class RichText extends Component {
+			let img;
+
+			if ( width > 150 ) {
+				img = `<img src="${ url }" alt="${ alt }" width="150" />`;
```
Shouldn't this be defined somewhere else (by the block/token)?

jayshenk replied:
Agreed. There has been some discussion on building an API for Inline Blocks in https://github.com/WordPress/gutenberg/pull/5794#issuecomment-377117224 and https://github.com/WordPress/gutenberg/pull/5794#issuecomment-377912079 but I haven't gotten to that yet. Would be great to get some thoughts on how best to approach that.

jayshenk replied:
https://github.com/WordPress/gutenberg/pull/5794/commits/06ce12556d7bfae783db4a857edca002ec1b91e5 adds a module for inline blocks and moves this logic there.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -441,6 +460,60 @@ export class RichText extends Component {
+		}
+
+		if ( isSelected ) {
+			// setTimeout prevents bug when switching between two
```
Which bug?

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -711,9 +784,20 @@ export class RichText extends Component {
+		}
+
+		if ( this.props.inlineBlock && this.props.isSelected ) {
+			this.insertInlineBlock();
```
Hm, I don't understand the `inlineBlock` prop here. It doesn't sound like it would trigger an action.

jayshenk replied:
Good point, I renamed it to `inlineBlockForInsert` in https://github.com/WordPress/gutenberg/pull/5794/commits/d5b87664194cefe7e54b2fd1fc304eb91260d27a. If that is still not clear, please let me know. This prop gets the `inlineBlockForInsert` state from the editor store when it is added there by the media library in the inserter.

iseulde commented on editor/components/inserter/index.js:
```diff
@@ -54,34 +101,63 @@ class Inserter extends Component {
+							onClose();
+						};
+
+						if ( this.state.isInline ) {
```
This is all very image specific? Would it make sense to generalise? Maybe good for a first iteration though...

nb replied:
That’s a good question, @jayshenk – do you think we should aim to merge this before or adding other inline blocks like mentions?

jayshenk replied:
Good question. I have been studying the autocompleters, of which currently there are just two: the user autocomplete (@ mention) and the block autocomplete (slash command). Since it has been decided not to show the Inline Blocks menu when a paragraph (the default block) is empty, I think this excludes the possibility of having the slash command as an inline block. Also the default block is the only one in which the slash command can be used. That leaves the @ mention, which I think I could include in this PR.

I have been studying the autocomplete to see how involved this would be. Need to look into it more but I think it should be doable. Perhaps there could be an inline blocks module that exports both the inline image and @ mention, but I need to do some further planning to decide the best way to structure this. Perhaps there would be a `type` property of inline blocks, which could be `image` (or something more generic?) or `autocomplete`, with the possibility of adding more types in the future.

The autocomplete type would need to trigger the autocomplete to run, which currently is triggered by `keypress` I believe, but now would need to be triggered when selected in the inserter. We would also need to think about if we would want to make this extensible, allowing plugins to add new inline blocks.

jayshenk replied:
https://github.com/WordPress/gutenberg/pull/5794/commits/06ce12556d7bfae783db4a857edca002ec1b91e5 adds an inline blocks module, in preparation for adding other inline blocks like @ mentions.

iseulde commented on editor/components/inserter/index.js:
```diff
@@ -39,12 +52,46 @@ class Inserter extends Component {
+
+	isInsertingInline() {
+		const { selectedBlock, canInsertInline } = this.props;
+		const isEmptyParagraph = selectedBlock &&
```
I don't understand the paragraph check here. It's generalised to all `RichText` components?

iseulde replied:
If you still need this logic for some reason, `isUnmodifiedDefaultBlock` might work better here.

jayshenk replied:
An empty Paragraph is a special case where the Inline Blocks menu should not render (see https://github.com/WordPress/gutenberg/pull/5794#issuecomment-376059920).

I agree, using `isUnmodifiedDefaultBlock` is a more elegant solution. Made the change in https://github.com/WordPress/gutenberg/pull/5794/commits/6bebead9db10c48e6ed531ec82b10c521feb83af.

iseulde commented:
> First was a concern about using the term token or inlineToken in the code to name Inline Blocks. 'Token' is rather opaque and may not be easy for other developers to understand what they are. Perhaps they should be named inlineBlocks to match the UI. My only concern would be whether they would be confused with standard blocks which have a different API.

I do like "token" a lot because it is does not mention "block" at all. It's hard to think about mentions, tags, footnotes, date... as inline blocks. They are very different technically, and also have different UI that comes with it. You can't do with them what you can with normal blocks. Maybe there are better words than "token" though.

StaggerLeee commented:
Difficult one. I already see how people insert images before-after paragraph blocks (before-after text inside one block). Believing it is native and normal WP way to insert images.

But for everything there is first time and short few minutes education can solve it all. To have inline images and developers can shortly tell and show to Users where to find it, and it is not normal way to add images.

I was thinking something different. People tend to copy/paste a lot of text from other sources, or one Word documet. All of it ends in one Paragraph block and then it is hard to avoid giving them Inline Image insert option. They simply have to have it in this circumstances.

But if, maybe if. When cursor is inside Paragraph block to give Users option to split Paragraph Block to two ? Then they would not need Inline image option.
- 3 dots icon - Split Block from here

jayshenk committed:
Use Lodash `get` to safely access property

jayshenk committed:
Use `isUnmodifiedDefaultBlock` instead of empty Paragraph check

jayshenk committed:
Remove `inlineBlocksEnabled` prop

Allow users the option of inserting Inline Blocks in any RichText.

jayshenk commented:
From @iseulde:
> I do like "token" a lot because it is does not mention "block" at all. It's hard to think about mentions, tags, footnotes, date... as inline blocks. They are very different technically, and also have different UI that comes with it. You can't do with them what you can with normal blocks. Maybe there are better words than "token" though.

Yes, I agree we don't want to confuse inline blocks with normal blocks. Since there was previous feedback https://github.com/WordPress/gutenberg/pull/5794#issuecomment-381246641 that the term "token" was also confusing, I'll think about if there is another name we can use (and suggestions are welcome!). CC @nb since I think he had some thoughts about this.

jayshenk committed:
Rename `inlineBlock` prop to `inlineBlockForInsert`

jayshenk committed:
Update comment for clarity

gziolo commented on editor/components/inserter/inline-menu.js:
```diff
@@ -0,0 +1,29 @@
+					id="editor-inserter__separator-inline"
+					aria-hidden="true"
+				>
+					Inline Blocks
```
This text should be wrapped with `i18` helper and another one a few lines later, too.

jayshenk replied:
Thanks for catching that! Fixed in https://github.com/WordPress/gutenberg/pull/5794/commits/5275914f57857e2a6109989cb04c3159401985d9.

jayshenk committed:
Use i18 helper

jayshenk committed:
Move insert position to RichText state

Prevents insert position calculation from happening on every render of
the insertion point.

jayshenk committed:
CSS for Image and Gallery Block compatibility

jayshenk committed:
Style resize handles to match Image block

nb commented on test/e2e/specs/multi-block-selection.test.js:
```diff
@@ -19,6 +19,7 @@ describe( 'Multi-block selection', () => {
 		// Creating test blocks
 		await page.click( '.editor-default-block-appender' );
 		await page.keyboard.type( 'First Paragraph' );
+		await page.keyboard.press( 'Enter' );
```
Why didn't we need this line before?

jayshenk replied:
Before, the standard blocks menu would render when the menu was clicked in the next line. However, in this branch, the Inline Blocks menu would render, preventing the standard Image block from being selected on line 24. So we need to press Enter here, to leave the paragraph, so that the standard blocks menu will render and the normal image block can be selected.

jayshenk committed:
Fix failing e2e test

Now that the Quote block uses nested blocks, we press Enter to move to
the next block within the Quote before clicking the menu. This is
necessary in order to render the standard inserter menu instead of the
Inline Blocks menu.

jayshenk commented:
From @StaggerLeee:
> But if, maybe if. When cursor is inside Paragraph block to give Users option to split Paragraph Block to two ? Then they would not need Inline image option.
> 
> 3 dots icon - Split Block from here

Thanks for your feedback and suggestion! Currently, you can split a Paragraph block into two separate Paragraph blocks by placing the caret where you want to split and typing 'Enter'. But if you would like to discuss the addition of an icon for splitting a paragraph, perhaps you could create a separate issue for that?

> Then they would not need Inline image option.

In cases where the user wants an image inline with text, I think they would still need this option, as the standard Image block does not have an option for inline alignment.

jayshenk committed:
Add reducer tests

jayshenk committed:
Add tests for actions

jayshenk committed:
Add tests for selectors

jayshenk committed:
Add inline blocks module and reorganize code to use it

Aims to make it possible to add more inline blocks besides inline
images.

iseulde commented on blocks/index.js:
```diff
@@ -14,6 +14,7 @@ import './hooks';
 // and then stored as objects in state, from which it is then rendered for editing.
 export * from './api';
 export * from './autocompleters';
+export * from './inline-blocks';
```
Currently we have a `core-blocks` directory to the top level? Would it make sense to have this one at the top level too with a similar way to register inline blocks?

jayshenk replied:
I like that idea! Commit https://github.com/WordPress/gutenberg/pull/5794/commits/4403d200433ded292223d7c9bb849b285eabbce5 sets up `core-inlineblocks` and `inline-blocks` directories to mirror the `core-blocks` and `blocks` directories and adds a similar way to register inline blocks. It's a work in progress.

iseulde commented on blocks/inline-blocks/inline-image.js:
```diff
@@ -0,0 +1,17 @@
+		const imgWidth = width > 150 ? 150 : width;
+		const img = `<img style="width:${ imgWidth }px;" src="${ url }" alt="${ alt }" />`;
+
+		editor.insertContent( img );
```
Would it be possible to remove the editor dependency here?

iseulde replied:
With that, it would also be cool if the block author could return a React element.

jayshenk replied:
I removed the editor dependency as you suggested in https://github.com/WordPress/gutenberg/pull/5794/commits/b3f6218266ac54024ffb62cb70ca1eedb3d699c2. I agree, having the option of returning a React element would be cool -- I'll look into how best to do that.

iseulde commented on blocks/rich-text/README.md:
````diff
@@ -139,4 +139,4 @@ registerBlockType( /* ... */, {
 ```
-{% end %}
\ No newline at end of file
+{% end %}
````
_squints eyes to see a difference_

iseulde commented on blocks/inline-blocks/inline-image.js:
```diff
@@ -0,0 +1,17 @@
+ */
+import { __ } from '@wordpress/i18n';
+
+export default {
```
I like how this is similar to normal block registration! :) Maybe also `export const settings`?

jayshenk replied:
Great idea! I used your suggestion and export `name` and `settings` now, similar to standard blocks https://github.com/WordPress/gutenberg/pull/5794/commits/4403d200433ded292223d7c9bb849b285eabbce5.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -387,6 +400,19 @@ export class RichText extends Component {
+		// Find the parent "relative" or "absolute" positioned container
+		const findRelativeParent = ( node ) => {
+			const style = window.getComputedStyle( node );
+			if ( style.position === 'relative' || style.position === 'absolute' ) {
```
I'm wondering if we could do something like `!== 'static'` here. Totally optional, I know this is exciting code.

jayshenk replied:
That's a good idea! Not needed now though, since this function was removed in https://github.com/WordPress/gutenberg/commit/45221727c4a4283637968792f6a38700697f1a9f.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -387,6 +400,19 @@ export class RichText extends Component {
+			return findRelativeParent( node.parentNode );
+		};
+
+		return findRelativeParent( this.editor.getBody() );
```
Might it be possible to avoid the wrapped function by e.g. defining `node` with `this.editor.getBody()` if it is not defined? Or perhaps requiring it, don't know what would be better.

jayshenk replied:
Now that this function has been removed in https://github.com/WordPress/gutenberg/commit/45221727c4a4283637968792f6a38700697f1a9f we no longer need to worry about this 😃.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -387,6 +400,19 @@ export class RichText extends Component {
+	getContainerNode() {
+		// Find the parent "relative" or "absolute" positioned container
+		const findRelativeParent = ( node ) => {
+			const style = window.getComputedStyle( node );
```
As mentioned earlier, totally optional as it's existing code, but might be good to move this to the top as a browser dependency. Just a personal preference.

jayshenk replied:
That makes sense. Made the change in https://github.com/WordPress/gutenberg/pull/5794/commits/7adab327c2327d890fce2a3c33b978f323950f22.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -418,6 +436,50 @@ export class RichText extends Component {
+	setInsertPosition() {
+		const container = this.getContainerNode();
+		const containerStyle = window.getComputedStyle( container );
+		const marginLeft = get( containerStyle, 'margin-left', 0 );
```
Is this reliable? What if there are multiple containers setting margins? Any way we could avoid this? I merely want to avoid breakage if someone updates the CSS of another component.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -418,6 +436,50 @@ export class RichText extends Component {
+	insertInlineBlock() {
+		const { inlineBlockForInsert, completeInlineInsert } = this.props;
+
+		if ( inlineBlockForInsert.type === 'image' ) {
```
Any way to avoid inline block specific logic here?

jayshenk replied:
I think I will try moving all of the inline block logic out of the RichText `index.js` and into a separate `InlineBlocks` component that RichText will render. I'll see if I can move some of this logic to the registered inline block as well.

jayshenk replied:
I've now moved all of the inline block logic into a new `InlineBlocks` component in https://github.com/WordPress/gutenberg/pull/5794/commits/d4229ef3135d85d0079d1c6f4481deb1192ec6f0.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -837,6 +928,26 @@ export class RichText extends Component {
 					</div>
 				) }
+				{ isSelected && isInlineInsertionPointVisible &&
+					<InlineInsertionPoint
```
I have no idea here from a quick look, but what prevents us from positioning the parent element above, and keeping a reference to that, instead of calculating the positioned parent element?

jayshenk replied:
Good question! I had assumed since there was existing code that calculated the positioned parent, that it must be necessary. But looking at it now I can't find a reason for it. Commit https://github.com/WordPress/gutenberg/pull/5794/commits/4abeaf72c7eec9d5450969273070323771196b28 uses a `ref` instead, as you suggested.

I wonder if `getFocusPosition` can use the `ref` as well? Since it is existing code that is used to position the `FormatToolbar`, I will refrain from changing it there for now, in case there is a reason for it.

iseulde commented on blocks/rich-text/index.js:
```diff
@@ -418,6 +436,50 @@ export class RichText extends Component {
+		if ( inlineBlockForInsert.type === 'image' ) {
+			this.setState( { mediaLibraryOpen: true } );
+		} else {
+			inlineBlockForInsert.render( this.editor );
```
Is there any way we can avoid `toggleInsertAvailable` by doing some slot/fill magic? Maybe this is not possible, just wondering out loud. :)

iseulde commented on core-blocks/gallery/style.scss:
```diff
@@ -38,6 +38,10 @@
 			overflow: auto;
+
+			img {
+				display: inline;
```
Any way we could somehow have a "front-end" style sheet in the inline block registrations that gets included/imported here? Probably thinking too much of SCSS as JS haha. Would be cool if we could just do

```scss
figcaption {
  ...

  &inlineBlockStyles
}
```

or something. Ignore me I know nothing about SCSS 🙈

Maybe @jasmussen knows some magic here to separate the styles.

iseulde replied:
Ooooooh, something like this maybe? https://sass-lang.com/guide#topic-6

iseulde replied:
```scss
figcaption {
  ...

  @include inline-block-styles();
}
```

jasmussen replied:
Not entirely sure about the context for this code, but if the goal is to reuse bits of CSS then yes, I think a mixin can be the right way to do it. We already do this for button hover/focus styles: https://github.com/WordPress/gutenberg/blob/master/edit-post/assets/stylesheets/_mixins.scss#L129

iseulde replied:
Yes, these would be the styles from all inline blocks together. Not sure about the best solution here, maybe the current one is good for a first version.

jayshenk replied:
Interesting thoughts here. I like the idea of using a mixin, but not sure if it's possible to combine a group of stylesheets into a single mixin. Perhaps examining how standard blocks add styles would be instructive here. It doesn't look like they use mixins, but instead each have their own stylesheets ([Image block for example](https://github.com/WordPress/gutenberg/blob/master/core-blocks/image/editor.scss)). Blocks are also given a [unique className](https://github.com/WordPress/gutenberg/blob/master/docs/block-api.md#supports-optional) -- maybe inline blocks could have a className property as well.

iseulde replied:
I think this is fine to figure out later.

iseulde commented on editor/components/inserter/index.js:
```diff
@@ -39,15 +48,47 @@ class Inserter extends Component {
+		const { showInlineInsertionPoint, showInsertionPoint } = this.props;
+
+		if ( this.isInsertingInline() ) {
+			this.setState( { isInline: true } );
```
Why do we need local state here?

jayshenk replied:
Using local state here prevents the Inline Blocks menu from briefly rendering the standard blocks menu while closing, as described in the second bullet point of https://github.com/WordPress/gutenberg/pull/5794#issuecomment-376059920. Perhaps there is a way to solve that issue without using local state though -- I'll think about it.

iseulde commented on editor/components/inserter/inline-menu.js:
```diff
@@ -0,0 +1,41 @@
+	render() {
+		const inlineBlocks = getInlineBlocks();
+		const items = inlineBlocks.map( ( inlineBlock ) => (
+			pick( inlineBlock, [ 'id', 'title', 'icon' ] )
```
I don't understand why we need to pick properties here. 🤔 Seems like a waste of energy?

jayshenk replied:
You're right, that's unnecessary. Removed in https://github.com/WordPress/gutenberg/pull/5794/commits/f84cb62b0b604eae12a1a6a9aa8623de9e65764a.

iseulde commented on blocks/inline-blocks/inline-image.js:
```diff
@@ -0,0 +1,17 @@
+	icon: 'format-image',
+	render( { url, alt, width }, editor ) {
+		const imgWidth = width > 150 ? 150 : width;
+		const img = `<img style="width:${ imgWidth }px;" src="${ url }" alt="${ alt }" />`;
```
1. Oh now that I think of it, it would be good to include the ID class name here like we do in the image block so images are caught by the WordPress' display filter and made responsive. (You can check on the front-end to test this).

2. Why setting `style` instead of `width` and `height` attributes? Does this need a comment if there's a reason?

jayshenk replied:
Good points! This commit uses your suggestions: https://github.com/WordPress/gutenberg/pull/5794/commits/c928207a243c36fe33a1e6e86272518445229f39.

jayshenk committed:
Undo accidental change

My vim configuration wanted to add a newline here.

jayshenk committed:
Add image ID class and comment

Enables WordPress display filter to make image responsive.

jayshenk committed:
Remove unnecessary property picking

jayshenk committed:
Use ref instead of calculating positioned parent element

jayshenk committed:
Set `getComputedStyle` as a browser dependency

jasmussen commented on blocks/rich-text/style.scss:
```diff
@@ -78,6 +78,19 @@
 		line-height: $editor-line-height;
 	}
+
+	div.mce-resizehandle {
```
> Would be cool if it looks just like the image block resining handles?

It would be cool. It's not a huuuge necessity, but it would be a nice enhancement, in this or a separate PR. 

I haven't tested recently, but I could swear that I saw a commit fly by that mentioned this?

jayshenk replied:
Yes, I thought I had them looking identical in commit https://github.com/WordPress/gutenberg/pull/5794/commits/2bd82b75df1331a70327cd726eea6fead8594e36 @jasmussen, but I may have missed something. What was the difference you noticed @iseulde?

mtias replied:
Having them look different might actually be beneficial in communicating it's not a block. Maybe not entirely different, but white instead of blue.

jayshenk replied:
That makes sense. I tried out the white background in https://github.com/WordPress/gutenberg/pull/5794/commits/7fbaea2ebecc34354f75d05d5ba9cba1ccf218d0. Let me know what you think.

<img width="731" alt="screen shot 2018-05-03 at 8 24 07 pm" src="https://user-images.githubusercontent.com/10903492/39609055-f85db714-4f13-11e8-96e9-cd1ea7bb207a.png">

iseulde replied:
Looks good! If needed we can adjust later.

jayshenk committed:
Begin setting up inline blocks API

Aims to create an API similar to that of standard blocks, so that new
inline blocks can be registered.

jayshenk committed:
Remove unused `getContainerNode` function

No longer needed after rebasing, since we are using `ref` of the
container instead.

jayshenk commented on blocks/rich-text/index.js:
```diff
@@ -387,6 +400,19 @@ export class RichText extends Component {
 		this.context.onCreateUndoLevel();
 	}
 
+	getContainerNode() {
```
I removed this function in https://github.com/WordPress/gutenberg/pull/5794/commits/45221727c4a4283637968792f6a38700697f1a9f since we are using a `ref` for the container now. Thanks for making PR https://github.com/WordPress/gutenberg/pull/6518 to help that happen!

jayshenk committed:
Fix failing e2e test

Reverts f817d3d since Quote block no longer using nested blocks after
rebase.

jayshenk committed:
Remove editor dependency from inline image settings

jayshenk committed:
Remove unnecessary margin calculation for insertion point

After rebasing and using a `ref` for the positioned container, I'm no
longer experiencing the problem with positioning the insertion point in
the List block that the margin calculation addressed.

jayshenk commented on blocks/rich-text/index.js:
```diff
@@ -837,6 +928,26 @@ export class RichText extends Component {
+						style={ this.state.insertPosition }
+					/>
+				}
+				{ this.state.mediaLibraryOpen &&
```
I'm not familiar with slot/fill, but it does look cool! I'll study it and see if it could be applied here.

jayshenk replied:
For now, I've moved all of the inline block logic into a new `InlineBlocks` component in https://github.com/WordPress/gutenberg/pull/5794/commits/d4229ef3135d85d0079d1c6f4481deb1192ec6f0. Haven't used slot/fill yet, but I will look for opportunities to do so 😄

mtias commented on blocks/rich-text/index.js:
```diff
@@ -704,9 +757,31 @@ export class RichText extends Component {
+		}
+
+		if (
+			this.props.isInlineInsertionPointVisible &&
```
Kind of seems all this functionality could be packaged in a HOC `withInlineInsertion` or something to not burden RichText so much. Not sure.

jayshenk replied:
Yes, I agree that I shouldn't burden RichText so much. I've now moved all of the inline block logic into a new `InlineBlocks` component in https://github.com/WordPress/gutenberg/pull/5794/commits/d4229ef3135d85d0079d1c6f4481deb1192ec6f0.

jayshenk committed:
Move inline block logic from RichText into new InlineBlocks component

jayshenk committed:
Experiment with resize handle styles

- try white background
- remove outline and selection background

iseulde commented on blocks/rich-text/inline-blocks.js:
```diff
@@ -0,0 +1,135 @@
+		} = this.props;
+
+		if ( inlineBlockForInsert.type === 'image' ) {
+			this.openMediaLibrary();
```
I think this is still a work in progress, but just noting for now that this does not belong in `RichText`, this should move to the block registration file.

jayshenk replied:
@iseulde Starting to work on this, along with having the inline block return a React element as you suggested in https://github.com/WordPress/gutenberg/pull/5794#discussion_r185042816. Do you think this should be limited to a single React element or could it be a full component with lifecycle methods? For the inline image it seems like it may need to be a component in order to include the MediaUpload logic. But I'm not sure if TinyMCE content can include React components. I'll look at this more tomorrow, but would appreciate any thoughts you or others have.

iseulde replied:
Hm, I'm not quite sure. RichText shouldn't be involved here until the React Element/HTML is ready to be inserted. Maybe it can be mounted in the inserter?

jayshenk committed:
Add E2E test for adding inline blocks

For now just tests adding an Inline Image in a Paragraph.

jayshenk committed:
Cleanup after rebase

RichText directory was moved from blocks/ to editor/components when I rebased.
This commit moves the InlineInsertionPoint and InlineBlocks to the new
location as well.

jayshenk commented:
Since a lot has changed since I originally wrote the PR description, I've [updated the description](https://github.com/WordPress/gutenberg/pull/5794#issue-177338578) with details about the current **Technical Design**, what tasks I'm currently working on, and a new gif. I welcome any feedback on that. I also changed the title of the PR to reflect that it now includes an API for inline blocks. CC @nb.

SuperGeniusZeb commented:
Shouldn't `core-inlineblocks` be `core-inline-blocks` to match `inline-blocks`?

jayshenk committed:
Rename core-inlineblocks directory to core-inline-blocks

Previously I had used only a single dash, in preparation for adding to
webpack where only the first dash was being replaced by the
`camelCaseDash` function. However, now that this function has been
updated to replace all dashes, this is no longer a limitation.

jayshenk commented:
> Shouldn’t core-inlineblocks be core-inline-blocks to match inline-blocks?

@SuperGeniusZeb Agreed! I just renamed it in https://github.com/WordPress/gutenberg/pull/5794/commits/2e5fca4654281739309c9a98479d746159275c1a.

iseulde commented on editor/components/rich-text/index.js:
```diff
@@ -439,6 +442,18 @@ export class RichText extends Component {
 		};
 	}
 
+	getInsertPosition() {
```
Why is this not part of `InlineBlocks` if the `editor` instance is passed?

jayshenk replied:
I had left if there since the `containerRef` and the similar `getFocusPosition` function were in `RichText`, but I agree that it belongs with `InlineBlocks` now, and moved it in https://github.com/WordPress/gutenberg/pull/5794/commits/b8a225acf9479b2dc8ae18cb73feab5168b62d0c.

iseulde commented on editor/components/rich-text/index.js:
```diff
@@ -122,6 +123,8 @@ export class RichText extends Component {
 		this.onPaste = this.onPaste.bind( this );
 		this.onCreateUndoLevel = this.onCreateUndoLevel.bind( this );
 		this.setFocusedElement = this.setFocusedElement.bind( this );
+		this.getFocusPosition = this.getFocusPosition.bind( this );
```
Remnant? Cannot see this in this diff.

jayshenk replied:
Thanks for catching that. Removed in https://github.com/WordPress/gutenberg/pull/5794/commits/f4a4358dbfeba0e5a1770a1632a5c5d920ccca91.

iseulde commented on editor/components/rich-text/index.js:
```diff
@@ -723,9 +738,12 @@ export class RichText extends Component {
 	}
 
 	componentDidUpdate( prevProps ) {
+		if ( ! this.editor ) {
```
Looks like this can be moved back to minimise diff.

jayshenk replied:
Fixed in https://github.com/WordPress/gutenberg/pull/5794/commits/f4a4358dbfeba0e5a1770a1632a5c5d920ccca91

iseulde commented on editor/components/rich-text/inline-blocks.js:
```diff
@@ -0,0 +1,135 @@
+		// When moving between two different RichText with the keyboard, we need to
+		// make sure `setInsertAvailable` is called after `setInsertUnavailable`
+		// from previous RichText so that editor state is correct
+		setTimeout( this.props.setInsertAvailable );
```
This needs to be cleaned up. There's a HOC for this now.

iseulde commented on editor/components/rich-text/inline-blocks.js:
```diff
@@ -0,0 +1,135 @@
+		super( ...arguments );
+
+		this.insert = this.insert.bind( this );
+		this.onSelectMedia = this.onSelectMedia.bind( this );
```
As previously commented, I don't like the image specific logic here. Would be good to be part of the block registration somehow.

jayshenk replied:
Agreed. It's tricky because if the registered block returns a React component, RichText will render it to a string so it can't use React methods after being inserted. I've been contemplating whether it would be possible to use slot/fill or another method to render an inline block's "menu" (in this case the Media Library) either in the `Inserter` or `InlineBlocks` component, and then have the menu component return the element to be rendered to string on select, but this still doesn't seem ideal. 

Related: your https://github.com/WordPress/gutenberg/pull/5794#discussion_r187945949 
Also maybe related: the Edit toolbar https://github.com/WordPress/gutenberg/pull/5794#issuecomment-381516924 and how that could be rendered by React after insertion.

iseulde commented on core-inline-blocks/inline-image.js:
```diff
@@ -0,0 +1,24 @@
+		// set width in style attribute to prevent Block CSS from overriding it
+		const img = `<img class="wp-image-${ id }" style="width:${ imgWidth }px;" src="${ url }" alt="${ alt }" />`;
+
+		return img;
```
Would be interesting if this is a React component and then render to string in RichText? Not sure though. Cc @youknowriad @aduth regarding RichText state and formats.

jayshenk replied:
Agreed. @iseulde added `edit` and `save` functions that return React elements in https://github.com/WordPress/gutenberg/pull/6959/files#diff-157ef14c5b252d176bfb2fa95850763bR20. Happy to merge those changes if we want, but I think maybe further development is moving to https://github.com/WordPress/gutenberg/pull/6959 now.

iseulde commented:
When testing, it seems to work nicely!

1. The caret is slightly misplaced:

<img width="150" alt="screen shot 2018-05-14 at 15 34 56" src="https://user-images.githubusercontent.com/4710635/40000731-dfaa98be-578c-11e8-8f60-eb584c7ce661.png">

2. It would be really nice if the image can be moved by drag and drop (without breaking multi-selection). So probably this should only work within the same block.

3. I would be nice to have an edit option as mentioned in https://github.com/WordPress/gutenberg/pull/5794#issuecomment-381516924.

jayshenk commented:
Thanks for doing another round of code review and for testing this out @iseulde! I will get working on the issues you brought up.

jayshenk committed:
cleanup diff

jayshenk committed:
Use withSafeTimeout HOC for setTimeout

jayshenk committed:
Move `getInsertPosition` from RichText to InlineBlocks

jayshenk commented on core-inline-blocks/inline-image.js:
```diff
@@ -0,0 +1,24 @@
+	icon: 'format-image',
+
+	render( { id, url, alt, width } ) {
+		const imgWidth = width > 150 ? 150 : width;
```
Good call. I noticed @iseulde used `Math.min` in https://github.com/WordPress/gutenberg/pull/6959/files#diff-157ef14c5b252d176bfb2fa95850763bR39. Happy to make that change here but it looks like development of this feature is moving to that branch now?

iseulde replied:
Oh, I already changed this yes. 😄

aduth commented:
Thanks @jayshenk . I missed that development had moved to #6959. Should we close this pull request?

iseulde commented:
Superseded by #6959.
